File: projects_with_reviews/python_number-conversion_with_correct_review/number-conversion/number_conversion.py
from string import ascii_uppercase


class Format:
    ROMAN = "ROMAN"
    BINARY = "BINARY"
    OCTAL = "OCTAL"
    HEXADECIMAL = "HEXADECIMAL"


ALPHABET_VALUES = {str(ord(c) - 55): c for c in ascii_uppercase}
ROMAN = [
    (1000, "M"),
    (900, "CM"),
    (500, "D"),
    (400, "CD"),
    (100, "C"),
    (90, "XC"),
    (50, "L"),
    (40, "XL"),
    (10, "X"),
    (9, "IX"),
    (5, "V"),
    (4, "IV"),
    (1, "I"),
]


def decimal_to_any_format(num, format_type) -> str:
    if format_type not in (Format.ROMAN, Format.BINARY, Format.OCTAL, Format.HEXADECIMAL):
        return ""
    return {
        Format.ROMAN: lambda: decimal_to_roman(num),
        Format.BINARY: lambda: decimal_to_any_base(num, 2),
        Format.OCTAL: lambda: decimal_to_any_base(num, 8),
        Format.HEXADECIMAL: lambda: decimal_to_any_base(num, 8),
    }[format_type]()


def decimal_to_roman(number: int) -> str:
    """
    Given a integer, convert it to an roman numeral. The number
    must be between 0 and 4000 (both included). It will return an error message if the input is invalid.
    """
    if not 0 <= number <= 4000:
        return "Invalid input. Input must be greater than 0 and less than 4000"
    result = []
    for arabic, roman in ROMAN:
        (factor, number) = divmod(number, arabic)
        result.append(roman * factor)
    return "".join(result)


def decimal_to_any_base(num: int, base: int) -> str:
    """
    Convert a positive integer to another base as str.
    """
    if base in (0, 1) or base > 36:
        return "Invalid base. Base should be between 2 and 36."

    if num < 0:
        return "Input number must be positive."

    new_value = ""
    div = 0
    while div != 1:
        div, mod = divmod(num, base)
        if base >= 11 and 9 < mod < 36:
            actual_value = ALPHABET_VALUES[str(mod)]
        else:
            actual_value = str(mod)
        new_value += actual_value
        div = num // base
        num = div
        if div == 0:
            return str(new_value[::-1])
        elif div == 1:
            new_value += str(div)
            return str(new_value[::-1])

    return new_value[::-1]

File: projects_with_reviews/java_number-conversion_with_correct_review/number-conversion/src/main/java/org/crprojects/NumberConversion.java
package org.crprojects;

public class NumberConversion {

    private static final int[] ARABIAN_NUMBERS = new int[] {
            1000,
            900,
            500,
            400,
            100,
            90,
            50,
            40,
            10,
            9,
            5,
            4,
            1,
    };

    private static final String[] ROMAN_NUMBERS = new String[] {
            "M",
            "CM",
            "D",
            "CD",
            "C",
            "XC",
            "L",
            "XL",
            "X",
            "IX",
            "V",
            "IV",
            "I",
    };

    public enum Format {
        ROMAN,
        BINARY,
        OCTAL,
        HEXADECIMAL
    }

    /**
     * This method produces a String value of any given input decimal to Roman,
     * binary, octal or hexadecimal format.
     *
     * @param num integer of which we need the value in converted format
     * @return converted value into desired format as a string, or an error message
     * if the input is invalid
     */
    public static String decimalToAnyFormat(int num, Format format) {
        return switch (format) {
            case ROMAN -> decimalToRoman(num);
            case BINARY -> decimalToAnyBase(num, 2);
            case OCTAL -> decimalToAnyBase(num, 8);
            case HEXADECIMAL -> decimalToAnyBase(num, 8);
        };
    }

    /**
     * This method produces a String value of any given input decimal to Roman
     * format.
     *
     * @param input decimal of which we need the value in Roman format,
     *              must be greater than 0 and less than 4000
     * @return converted value into Roman format as a string, or an error message
     * if the input is invalid
     */
    private static String decimalToRoman(int input) {
        if (input < 0 || input > 4000) {
            return "Invalid input. Input must be greater than 0 and less than 4000";
        }

        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < ARABIAN_NUMBERS.length; i++) {
            int arabian = ARABIAN_NUMBERS[i];
            String roman = ROMAN_NUMBERS[i];
            int times = input / arabian;
            builder.append(roman.repeat(times));
            input -= times * arabian;
        }

        return builder.toString();
    }

    /**
     * This method produces a String value of any given input decimal to Roman
     * format.
     *
     * @param input decimal of which we need the value in base in String format
     * @param base base to which we need to convert the decimal to
     * @return string format of the converted value in the given base
     */
    private static String decimalToAnyBase(int input, int base) {
        StringBuilder builder = new StringBuilder();

        if (input < 0) {
            return "Input number must greater than 0.";
        }

        if (input == 0) {
            return "0";
        }

        while (input > 0){
            int modulo = input % base;
            boolean singleDigit = modulo <= 9;
            char character = (char) (singleDigit ? modulo + '0' : modulo - 10 + 'A');
            builder.append(character);
            input /= base;
        }

        return builder.reverse().toString();
    }
}
